Implemented Multitype

LOuroboros opened the pull request:
## Description
This PR implements the only Gen. 4 ability that was missing; Arceus' Multitype.
Because of the checks involved, it will not work if the user doesn't have the item_expansion and pokemon_expansion branches present in their branch.
[Video showcase](https://www.youtube.com/watch?v=bFCPMTrPQTg)

## **Discord contact info**
Lunos#4026


*Please be gentle.*

ExpoSeed commented:
LGTM but i feel like this would be more appropriate in PE instead of BE. Fortunately moving it is easy, just a simple `git cherry-pick`, but I want to know others' thoughts

LOuroboros commented:
> 
> 
> LGTM but i feel like this would be more appropriate in PE instead of BE. Fortunately moving it is easy, just a simple `git cherry-pick`, but I want to know others' thoughts

Sure. The only reason why I didn't submit it there, as I already told to Ultima, is that then Multitype will be nothing but 2 text strings in the code.

I wouldn't mind that though. It's funny and funny things are funny :eyes:

ExpoSeed commented on src/party_menu.c:
```diff
@@ -6355,3 +6360,86 @@ void IsLastMonThatKnowsSurf(void)
+
+void SetArceusForm(struct Pokemon *mon)
+{
+#if defined(ITEM_EXPANSION) && defined(POKEMON_EXPANSION)
```
```suggestion
#if defined(POKEMON_EXPANSION)
```

LOuroboros commented:
Ready for re-review :eyes:

LOuroboros commented:
Nevermind. This is not ready for re-review. Please don't look at it.

LOuroboros commented:
Alright, this is now ready for re-review. I apologize for the fuckup :eyes:

LOuroboros commented:
I've been working on some extra changes for this feature. Those changes can be seen in **[my multitypePR_dev branch](https://github.com/LOuroboros/pokeemerald/commits/multitypePR_dev)**.

The summary of the changes is:

- I moved the `switch` statements of `SetArceusForm` and `SetArceusFormPSS` to new functions of their own because it made things easier to read.

- I tweaked the `if` checks around because I was multichecking hold effects, abilities and stuff needlessly.

- Compared to the PR, now I'm calling the `UpdateSpeciesSpritePSS` function that I implemented from within `SetArceusFormPSS` instead of calling it from inside `SetCursorMonData`.

We've been discussing in the Discord server the idea of getting rid of the check for `HOLD_EFFECT_PLATE` and have the functions that get Arceus' current form check for items instead. That way the `switch` statements could safely default to `SPECIES_ARCEUS`, which is faster than hiding the `switch` behind an `if/else` as this PR does.

I need confirmation that the changes are good before pushing them to the branch whose commits I'm PR'ing here though.

AsparagusEduardo commented:
> I've been working on some extra changes for this feature. Those changes can be seen in **[my multitypePR_dev branch](https://github.com/LOuroboros/pokeemerald/commits/multitypePR_dev)**.
> 
> The summary of the changes is:
> 
> * I moved the `switch` statements of `SetArceusForm` and `SetArceusFormPSS` to new functions of their own because it made things easier to read.
> * I tweaked the `if` checks around because I was multichecking hold effects, abilities and stuff needlessly.
> * Compared to the PR, now I'm calling the `UpdateSpeciesSpritePSS` function that I implemented from within `SetArceusFormPSS` instead of calling it from inside `SetCursorMonData`.
> 
> We've also been discussing in the Discord server the idea of getting rid of the check for `HOLD_EFFECT_PLATE` and have the functions that get Arceus' current form check for items instead. That way the `switch` statements could safely default to `SPECIES_ARCEUS`, which is faster than hiding the `switch` behind an `if/else` as this PR does.
> 
> I need confirmation that the changes are good before pushing them to the branch whose commits I'm PR'ing here and before starting to work on that idea though.

Yeah, the changes look good. Please add them here.

LOuroboros commented:
Alright, pushed. I'll start to tweak the `switch` function to use items instead of `HOLD_EFFECT_PLATE` in a few moments.

LOuroboros commented:
Ready for re-re-review :eyes:

ExpoSeed commented on src/party_menu.c:
```diff
@@ -6355,3 +6361,69 @@ void IsLastMonThatKnowsSurf(void)
+
+    switch (item)
+    {
+        case ITEM_FLAME_PLATE:
+            return SPECIES_ARCEUS_FIRE;
+        case ITEM_SPLASH_PLATE:
+            return SPECIES_ARCEUS_WATER;
+        case ITEM_ZAP_PLATE:
+            return SPECIES_ARCEUS_ELECTRIC;
+        case ITEM_MEADOW_PLATE:
+            return SPECIES_ARCEUS_GRASS;
+        case ITEM_ICICLE_PLATE:
+            return SPECIES_ARCEUS_ICE;
+        case ITEM_FIST_PLATE:
+            return SPECIES_ARCEUS_FIGHTING;
+        case ITEM_TOXIC_PLATE:
+            return SPECIES_ARCEUS_POISON;
+        case ITEM_EARTH_PLATE:
+            return SPECIES_ARCEUS_GROUND;
+        case ITEM_SKY_PLATE:
+            return SPECIES_ARCEUS_FLYING;
+        case ITEM_MIND_PLATE:
+            return SPECIES_ARCEUS_PSYCHIC;
+        case ITEM_INSECT_PLATE:
+            return SPECIES_ARCEUS_BUG;
+        case ITEM_STONE_PLATE:
+            return SPECIES_ARCEUS_ROCK;
+        case ITEM_SPOOKY_PLATE:
+            return SPECIES_ARCEUS_GHOST;
+        case ITEM_DRACO_PLATE:
+            return SPECIES_ARCEUS_DRAGON;
+        case ITEM_DREAD_PLATE:
+            return SPECIES_ARCEUS_DARK;
+        case ITEM_IRON_PLATE:
+            return SPECIES_ARCEUS_STEEL;
+        case ITEM_PIXIE_PLATE:
+            return SPECIES_ARCEUS_FAIRY;
+        default:
+            return SPECIES_ARCEUS;
+    }
```
this switch case will not compile without item expansion

LOuroboros commented:
Ready for re-re-re-review 👀

ExpoSeed reviewed (approved):
i still believe this would be better suited in PE, but i won't stop this from being merged if BE is the majority opinion

LOuroboros commented:
> 
> 
> i still believe this would be better suited in PE, but i won't stop this from being merged if BE is the majority opinion

In the end, it doesn't really matter imo. Items, forms, the ability, everything is required to get the Arceus Forms going.
It's irrelevant, and if I have to choose one of the 3 branches to put this on, it would be the BE simply because otherwise, Multitype would be nothing but 2 text strings in the code.
At least this gives it some worth :eyes:

AsparagusEduardo commented:
> > i still believe this would be better suited in PE, but i won't stop this from being merged if BE is the majority opinion
> 
> In the end, it doesn't really matter imo. Items, ~forms~ species data, the ability, everything is required to get the Arceus Forms going.
> It's irrelevant, and if I have to choose one of the 3 branches to put this on, it would be the BE simply because otherwise, Multitype would be nothing but 2 text strings in the code.
> At least this gives it some worth 👀

I think I see where Expo's coming from though. This code should hopefully be reusable for stuff like Giratina and Genesect form changing, so having all three on Pokémon Expansion would make sense.

ExpoSeed commented:
> Multitype would be nothing but 2 text strings in the code. At least this gives it some worth
well in PE, it would be worth about the same since you'd ifdef the multitype check there. also i disagree about the idea of giving worth for the sake of giving worth, it's still just as meaningless

> In the end, it doesn't really matter imo... It's irrelevant
I disagree, because otherwise it becomes really confusing what goes in which branch, and then juggling the three branches and their interdependencies becomes even more hellish than it already is. also, it would make it harder to create a broader system that generalizes across multiple pokemon. 

for instance, silvally, genesect, and arceus work near identically, but since genesect's form changes don't depend on ability, it would be placed into PE/IE, and then it becomes impossible to reuse the form change code for both.

taking this further, gracidea and shaymin are in item expansion, so it becomes impossible to unite them into one table with a form change field (similar to how it's done for evolutions).

as a result, i think paying careful attention to what goes where and staying consistent is critical

LOuroboros commented:
> > > i still believe this would be better suited in PE, but i won't stop this from being merged if BE is the majority opinion
> > 
> > 
> > In the end, it doesn't really matter imo. Items, ~forms~ species data, the ability, everything is required to get the Arceus Forms going.
> > It's irrelevant, and if I have to choose one of the 3 branches to put this on, it would be the BE simply because otherwise, Multitype would be nothing but 2 text strings in the code.
> > At least this gives it some worth 👀
> 
> I think I see where Expo's coming from though. This code should hopefully be reusable for stuff like Giratina and Genesect form changing, so having all three on Pokémon Expansion would make sense.

Giratina's and Genesect's are explicitly item based though, so those should go in the IE.

> > Multitype would be nothing but 2 text strings in the code. At least this gives it some worth
> 
> well in PE, it would be worth about the same since you'd ifdef the multitype check there. also i disagree about the idea of giving worth for the sake of giving worth, it's still just as meaningless

I don't see the problem with it myself.

> > In the end, it doesn't really matter imo... It's irrelevant
> 
> I disagree, because otherwise it becomes really confusing what goes in which branch, and then juggling the three branches and their interdependencies becomes even more hellish than it already is. also, it would make it harder to create a broader system that generalizes across multiple pokemon.
> 
> for instance, silvally, genesect, and arceus work near identically, but since genesect's form changes don't depend on ability, it would be placed into PE/IE, and then it becomes impossible to reuse the form change code for both.
> 
> taking this further, gracidea and shaymin are in item expansion, so it becomes impossible to unite them into one table with a form change field (similar to how it's done for evolutions).
> 
> as a result, i think paying careful attention to what goes where and staying consistent is critical, especially if we still want to keep the branches separate

I really fail to see this confusion or mayhem you're talking about, not gonna lie.

Forms will heavily rely on the contents scattered throughout the 3 branches no matter the case, so locking a hypothetical table as such behind an `#if defined (BATTLE_ENGINE) && defined (ITEM_EXPANSION)` if one ended up being implemented in the PE would be required regardless, wouldn't it?

As such, as long as the content is implemented in a consistent manner for each of the branches, nothing else should matter imo.

The form change of Arceus relies on the ability as much as it relies on the Plates, so in my opinion it would most definitely not go in the PE branch at the end of the day.

ExpoSeed commented:
> would be required regardless, wouldn't it?

Only for the entry in the table (again, just like evolutions). It wouldn't be needed in the code

> The form change of Arceus relies on the ability as much as it relies on the Plates

The ability never changes as part of the overall change, but the species does, and species is the domain of PE. Also there's only a single check for multitype, while for species there is both the switch case and the range check

LOuroboros commented:
> > would be required regardless, wouldn't it?
> 
> Only for the entry in the table (again, just like evolutions). It wouldn't be needed in the code

Hmm... do you have a quick example as to how it would be laid out?

I really don't see it myself. If the actual code doesn't need to be locked inside an ifdef, then... what is the cause of concern here?
 
> > The form change of Arceus relies on the ability as much as it relies on the Plates
> 
> The ability never changes as part of the overall change, but the species does, and species is the domain of PE. Also there's only a single check for multitype, while for species there is both the switch case and the range check

The `switch` doesn't check for species, it sets the species. The range check that matters contains the check for Multitype.

Normally, Multitype would be in charge of changing Arceus' type. Since we're not doing actual forms here but a species swap instead, the effect is no different.

Multitype is directly required to make the form change happen at all, and abilities are BE territory.

ExpoSeed commented:
> Hmm... do you have a quick example as to how it would be laid out?
> 
> I really don't see it myself. If the actual code doesn't need to be locked inside an ifdef, then... what is the cause of concern here?

of course
```c
struct FormChange
{
    u16 method;
    u16 param1;
    u16 param2;
    u16 targetSpecies;
};
...
const struct FormChange gFormChanges[][] = {
#if defined(ITEM_EXPANSION) && defined(ABILITY_MULTITYPE)
    [SPECIES_ARCEUS] = {{FORMCHANGE_HOLD_ITEM, ABILITY_MULTITYPE, ITEM_FLAME_PLATE, SPECIES_ARCEUS_FIRE},
                                       {FORMCHANGE_HOLD_ITEM, ABILITY_MULTITYPE, ITEM_SPLASH_PLATE, SPECIES_ARCEUS_WATER},
...
#endif
}
```
notice the similarities between this and the evolution table. with this, the data and the inputs to form change functions depend on branch, so they get ifdefs. however, the logic behind a form change does not depend on every branch, so there are no need for ifdefs

> The switch doesn't check for species

i never said it did, i'm just saying that arceus form changes manipulate and use the species much more than they do the ability, which is an indicator of what is more closely tied to the form change

LOuroboros commented:
> 
> 
> > Hmm... do you have a quick example as to how it would be laid out?
> > I really don't see it myself. If the actual code doesn't need to be locked inside an ifdef, then... what is the cause of concern here?
> 
> of course
> 
> ```c
> struct FormChange
> {
>     u16 method;
>     u16 param1;
>     u16 param2;
>     u16 targetSpecies;
> };
> ...
> const struct FormChange gFormChanges[][] = {
> #if defined(ITEM_EXPANSION) && defined(ABILITY_MULTITYPE)
>     [SPECIES_ARCEUS] = {{FORMCHANGE_HOLD_ITEM, ABILITY_MULTITYPE, ITEM_FLAME_PLATE, SPECIES_ARCEUS_FIRE},
>                         {FORMCHANGE_HOLD_ITEM, ABILITY_MULTITYPE, ITEM_SPLASH_PLATE, SPECIES_ARCEUS_WATER},
> ...
> #endif
> }
> ```
> 
> notice the similarities between this and the evolution table. with this, the data and the inputs to form change functions depend on branch, so they get ifdefs. however, the logic behind a form change does not depend on every branch, so there are no need for ifdefs

I mean, the logicstics matter. If certain form changes do end up going in the IE like they should do due to their nature, you will need the ifdefs anyway.

For the form changes to actually happen, you *will* have to require the ITEM_EXPANSION branch for sure.

I still don't agree with the idea and I still think Multitype is vital to the effect as without it, the form change can not happen.

The form change may not necessarily be battle related, but ability effects are.

> > The switch doesn't check for species
> 
> i never said it did, i'm just saying that arceus form changes manipulate and use the species much more than they do the ability, which is an indicator of what is more closely tied to the form change

The thing is that what manipulates the current species is a function that can only be triggered if the species if Arceus and its ability is Multitype. Both checks are integral.

Having one extra check that solely relies on the species matters not, when the function that is triggered by it won't be executed if the species' current ability is not Multitype.

You said something that I didn't really address earlier: "*Also there's only a single check for multitype, while for species there is both the switch case and the range check*"

The checks where Multitype is specifically involved are those that decide whether a form change can happen at all or not. If the species does not have the ability, they don't happen because they can't and that's not something you can just brush off imo. Effects where an ability is directly involved for an effect to happen are better suited in the BE, from my point of view.

I still reeeeally don't like this at all, but if there's more approval to your suggestion, I'll have to end up moving it to the PE whether I like it or not, so I'll be pinging some people to discuss this further.

@ultima-soul, @AsparagusEduardo, @Pyredrid, @MandL27, @ghoulslash, I'd like to hear your opinions too.

If you guys agree, I will close this PR and move the code to the PE.

AsparagusEduardo commented:
> The thing is that what manipulates the current species is a function that can only be triggered if the species if Arceus and its ability is Multitype. Both checks are integral.
> 
> Having one extra check that solely relies on the species matters not, when the function that is triggered by it won't be executed if the species' current ability is not Multitype.
> 
> You said something that I didn't really address earlier: "_Also there's only a single check for multitype, while for species there is both the switch case and the range check_"
> 
> The checks where Multitype is specifically involved are those that decide whether a form change can happen at all or not. If the species does not have the ability, they don't happen because they can't and that's not something you can just brush off imo. Effects where an ability is directly involved for an effect to happen are better suited in the BE, from my point of view.
> 
> I still reeeeally don't like this at all, but if there's more approval to your suggestion, I'll have to end up moving it to the PE whether I like it or not, so I'll be pinging some people to discuss this further.
> 
> @ultima-soul, @AsparagusEduardo, @Pyredrid, @MandL27, @ghoulslash, I'd like to hear your opinions too.
> 
> If you guys agree, I will close this PR and move the code to the PE.

I think it would be better to have the code in PE despite having the ability check, as it can be useful for applying it to Genesect and others, as it would be cleaner to have it only once instead of both PE and BE.

ExpoSeed commented:
> I mean, the logicstics matter. If certain form changes do end up going in the IE like they should do due to their nature, you will need the ifdefs anyway.

well yeah, but i was just planning on something like this
```c
void ItemUseCB_FormChangeItem(u8 taskId, TaskFunc task)
{
#ifdef POKEMON_EXPANSION
    TryFormChange(whatever) // this function would be in PE
#else
    gPartyMenuUseExitCallback = FALSE;
    PlaySE(SE_SELECT);
    DisplayPartyMenuMessage(gText_WontHaveEffect, TRUE);
    ScheduleBgCopyTilemapToVram(2);
    gTasks[taskId].func = task;
    return;
#endif
}
```
which handles that issue

> I still don't agree with the idea and I still think Multitype is vital to the effect as without it, the form change can not happen.

so is the pokemon? without the pokemon arceus, the form change cannot happen. i think this is true of all form changes of this nature (i.e. genesect does not have an ability to depend on)

> Both checks are integral.
agreed, but not for genesect, zacian, and zamazenta. having code in BE makes it impossible to reuse for these three

> If the species does not have the ability, they don't happen because they can't and that's not something you can just brush off imo.

the check also fails if a pokemon has the ability but is the wrong species. furthermore, once the ability check succeeds (or doesn't happen in the case of genesect), the ability is never referred to again, while the species gets further manipulated later.

> Effects where an ability is directly involved for an effect to happen are better suited in the BE, from my point of view.

my argument was that held item form changes are an extension of the pokemon and not an extension of the ability so they should be placed in PE. not only would this be more consistent (especially with how we handle evolutions), it would make the code much much much much easier to reuse for multiple pokemon (which cannot happen right now because code is fragmented across multiple branches)

LOuroboros commented:
> 
> 
> > I mean, the logicstics matter. If certain form changes do end up going in the IE like they should do due to their nature, you will need the ifdefs anyway.
> 
> well yeah, but i was just planning on something like this
> 
> ```c
> void ItemUseCB_FormChangeItem(u8 taskId, TaskFunc task)
> {
> #ifdef POKEMON_EXPANSION
>     TryFormChange(whatever) // this function would be in PE
> #else
>     gPartyMenuUseExitCallback = FALSE;
>     PlaySE(SE_SELECT);
>     DisplayPartyMenuMessage(gText_WontHaveEffect, TRUE);
>     ScheduleBgCopyTilemapToVram(2);
>     gTasks[taskId].func = task;
>     return;
> #endif
> }
> ```
> 
> which handles that issue. the further logic does not need any ifdefs

Alright, and `TryFormChange` which would encompass the Arceus form change and many others wouldn't need to strictly rely on the BATTLE_ENGINE since Multitype is already defined in the PE.

I can see the value in it. I don't like it, but I agree it may be for the better to put the Arceus forms there for convenience sake.

> > I still don't agree with the idea and I still think Multitype is vital to the effect as without it, the form change can not happen.
> 
> so is the pokemon? without the pokemon arceus, the form change cannot happen. i think this is true of all form changes of this nature (i.e. genesect does not have an ability to depend on)

No, but it does have an item to rely on which is why it'll go in the IE, as far as it's been discussed.

> > Both checks are integral.
> > agreed, but not for genesect, zacian, and zamazenta. having code in BE makes it impossible to reuse for these three
> 
> > If the species does not have the ability, they don't happen because they can't and that's not something you can just brush off imo.
> 
> the check also fails if a pokemon has the ability but is the wrong species. furthermore, once the ability check succeeds (or doesn't happen in the case of genesect), the ability is never referred to again, while the species gets further manipulated later.

Fair enough.

> > Effects where an ability is directly involved for an effect to happen are better suited in the BE, from my point of view.
> 
> my argument was that held item form changes are an extension of the pokemon and not an extension of the ability so they should be placed in PE. not only would this be more consistent (especially with how we handle evolutions), it would make the code much much much much easier to reuse for multiple pokemon (which cannot happen right now because code is fragmented across multiple branches)

That I cannot really agree with. From my point of view, held item form changes would be an extension of the items themselves. They're the ones causing an effect on the species.

I would never put any of the strictly item based form changes anywhere but in the IE. Me putting Arceus' forms in the BE is solely because Multitype is directly involved in the effect, and I don't think that should go in the PE myself.

ExpoSeed commented:
> Alright, and TryFormChange which would encompass the Arceus form change and many others wouldn't need to strictly rely on the BATTLE_ENGINE since Multitype is already defined in the PE.

yep that's the idea, we can reuse the same chunk of code for lots of pokemon

> No, but it does have an item to rely on which is why it'll go in the IE, as far as it's been discussed.

if it's in the IE it cannot reuse code with arceus since they work fundamentally the same, just different items and different abilities

> From my point of view, held item form changes would be an extension of the items themselves.

fair enough, but i think since the common factor is the pokemon, and the main thing that changes is the pokemon, i would put them in the PE (just like how `EVO_ITEM_MALE` and `EVO_ITEM_FEMALE` are grouped with the pokemon, not with the item)

LOuroboros commented:
> if it's in the IE it cannot reuse code with arceus even though they work fundamentally the same, just different items and different abilities

No, but it can reuse code with Giratina, which will also go in the IE as far as I understand it.

> fair enough, but i think since the common factor is the pokemon, and the main thing that changes is the pokemon, i would put them in the PE (just like how EVO_ITEM_MALE and EVO_ITEM_FEMALE are grouped with the pokemon, not with the item)

The thing is evolution methods are directly tired to the species and happen because of them too.

Genesect's is a situation where the species is directly affected by the item it's holding, which is really the exact same situation for Arceus.

ExpoSeed commented:
> No, but it can reuse code with Giratina, which will also go in the IE as far as I understand it.

if it was in pe, it would work for giratina, arceus, silvally, genesect, and giratina, rather than duplicating code twice

> The thing is evolution methods are directly tired to the species and happen because of them too.
> 
> Genesect's is a situation where the species is directly affected by the item it's holding, which is really the exact same situation for Arceus.

i don't understand, what is the difference between the evolution method and the form change? both happen because of the species and both check for items, which indicates to me that both should be treated roughly the same way.

LOuroboros commented:
> if it was in pe, it would work for giratina, arceus, silvally, genesect, and giratina, rather than duplicating code twice

Nothing to add to what I already said. I don't like the idea myself, but I'm not against moving the Arceus form change to the PE for convenience sake.

> i don't understand, what is the difference between the evolution method and the form change? both happen because of the species and both check for items, which indicates to me that both should be treated roughly the same way.

The change doesn't happen because of the species, it happens because of the item mainly. It's the item that triggers an effect on the species, not the other way around or anything like that.

Genesect changes explicitly and only because of the item it's holding. And the same happens with Arceus, except it normally changes its type alone. Arceus' type changes because it has the Multitype ability and the type is decided by the Plate it's holding.

That's IE, and BE/IE territory respectively.

Meanwhile, the evolution methods that you provided as an example happen because of the species being exposed to an item. The item is triggering a reaction in their bodies that produces the evolution.

Honestly, I wouldn't be against moving those to the IE, but evolutions are a fundamental part of the species and their data which is why they go in the PE.

Meanwhile, the form changes are alterations in a Pokémon's organism that are both triggered and dependant on other factors. Again, in the case of Genesect it's the Drives, and in the case of Arceus is its innate ability and the Plates.

I don't think I have anything else to add myself, specially after declaring that I am willing to move the stuff to either the IE or the PE if the other reviewers deem it a good idea. I explained my reasoning to introduce them in the BE and my stance on the matter, and I'm content with it, but if a majority comes to a decision, I will gladly abide to whatever is decided :eyes:

The rest depends on Pyredrid, Mandl, Ultima, Ghoulslash and Eduardo.

ExpoSeed commented:
> The change doesn't happen because of the species, it happens because of the item mainly. It's the item that triggers an effect on the species, not the other way around or anything like that.

the new evolution methods for items is implemented in PE, not IE

> Meanwhile, the evolution methods that you provided as an example happen because of the species being exposed to an item. The item is triggering a reaction in their bodies that produces the evolution.

i think it's the same for form changes. the species starts holding an item, then the species changes

> I don't think I have anything else to add myself, specially after declaring that I am willing to move the stuff to either the IE or the PE if the other reviewers deem it a good idea. I explained my reasoning to introduce them in the BE and my stance on the matter, and I'm content with it, but if a majority comes to a decision, I will gladly abide to whatever is decided 👀

sure, i think we'll have to agree to disagree

to summarize, my view is that all form changes should be implemented in PE. there are a number of parallels between form changes and evolutions (both take an input species, possibly check for a specified item, and turn the species into a target species). given these parallels, it is possible to craft code based on this PR that will generalize to multiple pokemon (genesect, giratina, silvally, arceus, etc.) use the same logic, making it easy to refactor for multiple pokemon at once and trivial to add new form changes. this can only happen if they are all in the same branch (instead of genesect and giratina being in IE while silvally and arceus are in BE). since in this situation, form changes are just an extension of pokemon data, it should be defined in PE.

that's my position on the matter, and even though i'll support the majority decision (evidenced by my approval of this PR), i do caution that fragmenting form change code across branches would make it harder to maintain and more unfriendly to use

LOuroboros commented:
> 
> 
> > The change doesn't happen because of the species, it happens because of the item mainly. It's the item that triggers an effect on the species, not the other way around or anything like that.
> 
> the new evolution methods for items is implemented in PE, not IE

Yes, as I said, the items cause a reaction in the Pokémon's body that triggers the evolution, just like I also said:

> Honestly, I wouldn't be against moving those to the IE, but evolutions are a fundamental part of the species and their data which is why they go in the PE.


> > Meanwhile, the evolution methods that you provided as an example happen because of the species being exposed to an item. The item is triggering a reaction in their bodies that produces the evolution.
> 
> i think it's the same for form changes. the species starts holding an item, then the species changes

The thing is that it then goes back to normal if you take it away from it. You see the difference here? The item is both, in charge of making the change happen and in charge of mantaining it as well.

> > I don't think I have anything else to add myself, specially after declaring that I am willing to move the stuff to either the IE or the PE if the other reviewers deem it a good idea. I explained my reasoning to introduce them in the BE and my stance on the matter, and I'm content with it, but if a majority comes to a decision, I will gladly abide to whatever is decided 👀
> 
> sure, i think we'll have to agree to disagree
> 
> to summarize, my view is that all form changes should be implemented in PE. there are a number of parallels between form changes and evolutions (both take an input species, possibly check for a specified item, and turn the species into a target species). given these parallels, it is possible to craft code based on this PR that will generalize to multiple pokemon (genesect, giratina, silvally, arceus, etc.) use the same logic, making it easy to refactor for multiple pokemon at once and trivial to add new form changes. this can only happen if they are all in the same branch (instead of genesect and giratina being in IE while silvally and arceus are in BE). since in this situation, form changes are just an extension of pokemon data, it should be defined in PE.
> 
> that's my position on the matter, and even though i'll support the majority decision (evidenced by my approval of this PR), i do caution that fragmenting form change code across branches would make it harder to maintain and more unfriendly to use

Agreed.

ExpoSeed commented:
On hold until we decide form table shenanigans and branch placement
